### ISSUE TITLE
Problem visualizing cam on trained model

### ISSUE THREAD

**Issue by davcaste** [jacobgil/pytorch-grad-cam#189]
Hi, 
I am using this script to evaluate my results on a brand classification for cars.
When I run this algoithm on the model in pytorch library (models.resnet34) pretrained on imagenet in which i just changed the classification head with:

`    
input_size = model_resnet.fc.in_features

  model_resnet.fc = nn.Sequential(
      nn.Linear(input_size, 256),
      nn.ReLU(),
      nn.Dropout(0.2),
      nn.Linear(256, num_classes),
      nn.Softmax(),
  )
`
the cam that I got as output actually make sense and look like this
![gradcam_cam](https://user-images.githubusercontent.com/55695300/147931769-a6c0e1c0-74a5-4cbb-843f-d3bca1fa4e53.jpg)
![gradcam_cam_gb](https://user-images.githubusercontent.com/55695300/147931822-3af00c40-02f7-4ff6-8fc5-f7a6ffc68a22.jpg)

While when i load the resnet34 with the new head (same structure as before), that I trained from scratch and that has an accuracy > 90% for each class, it gives me an activation map that doens't make sense and it's the same no matter what input image (given that it belongs to the same class).
![gradcam_cam (1)](https://user-images.githubusercontent.com/55695300/147932444-56a43d23-4728-475b-9a65-4e5a8dc7179a.jpg)
![gradcam_cam_gb (1)](https://user-images.githubusercontent.com/55695300/147932453-245deaca-659d-4c63-bd37-4e2d363165cd.jpg)
I'm struggling to come up to an explanation but I don't understand it because the perfromance doesn't match the cam.
I would be very gratefull if you have some advice.

**Comment by jacobgil** [comment]
Hi,
Some questions:
1. In case of the new pretrained network, is the normalization the same as before? Or maybe the new image needs to be normalized differently (0.5 instead of imagenet normalization etc).
2. Can you try removing the softmax from the network?
3. Try changing grad-cam to another method like AblationCAM and see if it's still the same?

**Comment by davcaste** [comment]
Thank you for the fast response.
Regarding your questions:
1. Yes I'm using the normalization with mean and std of imagenet: transforms.Normalize([0.485, 0.456, 406], [0.229, 0.224, 0.225]), the only different preprocessing that i made is to apply a new resize_padding that resize the bigger side of the image to 512 and rescale the smaller one maintaining the proportions and pad the difference with 0 to obtain a 512x512 image.
2. Removing the Softmax i obtain an gb image that is not completely grey but it doesn't solve the fact that i obtain the same cam no matter what input i give.
3. Using ablationCAM seems not to change the fact that i always get the same activation map

**Comment by jacobgil** [comment]
Lets try to diagnose if there is a problem with the CAM part, or in loading the model / processing the images: 

Can you try running the network on a few test images,
without using CAM?
Are the predictions correct for different categories? Do they make sense? Are they constant or do they change?

Is the model in eval?

**Comment by davcaste** [comment]
I thought that the problem was for the preprocessing because maybe the images of the same class has similar lenght/width ratio so the model specializes in recognizing the black stripes, anyway even removing the resize_padding in the testing part i obtain quite good results and the brands are predicted correctly most of the times.

Yes the model it's in eval.

The thing that is harder to understand to me is why the model pretrained on imagenet actually has a good activation map while the one that I trained that is actually able to recognize the brands have this kind of constant activation map

**Comment by jacobgil** [comment]
Hi, so just to make sure after removing resize_padding the CAMs look good?
Or currently they still look bad?

**Comment by jacobgil** [comment]
When you do resize_padding, we have to take care with visualizing the CAM since the tensor has different dimensions than the actual image.
So we should take the center part of the CAM, and then resize it back to the original tensor.
Is that currently happening?
Maybe that explains the effect with the constant CAM?

**Comment by davcaste** [comment]
I'll show you what I'm doing on the images with my code:

```
    image = Image.open(io.BytesIO(resp))
    im = image.convert("RGB")
    rgb_img = np.array(im)
    rgb_img = np.float32(rgb_img) / 255
    input_tensor = preprocess_image(im,
                                    mean=[0.485, 0.456, 0.406],
                                    std=[0.229, 0.224, 0.225])


```
```
def preprocess_image(img: np.ndarray, mean=[0.5, 0.5, 0.5], std=[0.5, 0.5, 0.5]) -> torch.Tensor:
    preprocessing = Compose([
        NewPad(image_size=512),
        ToTensor(),
        Normalize(mean=mean, std=std)
    ])
    new_img = preprocessing(img.copy())
    return new_img.unsqueeze(0)
```
then i compute the greyscale cam on this tensor as in the original code,

and finally the only function that i change is:

```
def show_cam_on_image(img: np.ndarray,
                      mask: np.ndarray,
                      use_rgb: bool = False,
                      colormap: int = cv2.COLORMAP_JET) -> np.ndarray:
    """ This function overlays the cam mask on the image as an heatmap.
    By default the heatmap is in BGR format.

    :param img: The base image in RGB or BGR format.
    :param mask: The cam mask.
    :param use_rgb: Whether to use an RGB or BGR heatmap, this should be set to True if 'img' is in RGB format.
    :param colormap: The OpenCV colormap to be used.
    :returns: The default image with the cam overlay.
    """
    heatmap = cv2.applyColorMap(np.uint8(255 * mask), colormap)
    if use_rgb:
        heatmap = cv2.cvtColor(heatmap, cv2.COLOR_BGR2RGB)
    heatmap = np.float32(heatmap) / 255
    
    if np.max(img) > 1:
        raise Exception(
            "The input image should np.float32 in the range [0, 1]")
        
    pad = NewPad(image_size=512)
    im = Image.fromarray(np.uint8((img)*255))
    img = pad(im)
    img = np.array(img)/255
    cam = heatmap + img
    cam = cam / np.max(cam)
    return np.uint8(255 * cam)
```

in order to adapt the dimension of the image to the one of the mask.

For completeness i also add the function that i use for the resize_padding:

```
def resize(image, img_size):
    w, h = image.size
    max_wh = np.max([w, h])
    ratio = max_wh / img_size
    new_w = int(w / ratio)
    new_h = int(h / ratio)
    image = transforms.Resize(size=(new_h, new_w))(image)
    return image

def get_padding(image):    
    w, h = image.size
    max_wh = np.max([w, h])
    h_padding = (max_wh - w) / 2
    v_padding = (max_wh - h) / 2
    l_pad = h_padding if h_padding % 1 == 0 else h_padding+0.5
    t_pad = v_padding if v_padding % 1 == 0 else v_padding+0.5
    r_pad = h_padding if h_padding % 1 == 0 else h_padding-0.5
    b_pad = v_padding if v_padding % 1 == 0 else v_padding-0.5
    padding = (int(l_pad), int(t_pad), int(r_pad), int(b_pad))
    return padding

class NewPad(object):
    def __init__(self, image_size, fill=0, padding_mode='constant'):
        assert isinstance(fill, (numbers.Number, str, tuple))
        assert padding_mode in ['constant', 'edge', 'reflect', 'symmetric']
        
        self.img_size = image_size
        self.fill = fill
        self.padding_mode = padding_mode
        
    def __call__(self, img):
        """
        Args:
            img (PIL Image): Image to be padded.

        Returns:
            PIL Image: Padded image.
        """
        img = resize(img, self.img_size)
        img = pad(img, get_padding(img), self.fill, self.padding_mode)
        return img
    
    def __repr__(self):
        return self.__class__.__name__ + '(padding={0}, fill={1}, padding_mode={2})'.\
            format(self.fill, self.padding_mode)
```

By doing what i mentioned above i obtain this cam for the class with higher score for every input image (but the model classify them correctly):
![gradcam_cam (2)](https://user-images.githubusercontent.com/55695300/147956612-619fdf2f-7861-4a75-8e2d-a8336cdbc186.jpg)
![gradcam_cam_gb (2)](https://user-images.githubusercontent.com/55695300/147956621-818916cd-417e-414d-b585-9ed9d06c9306.jpg)

**Comment by davcaste** [comment]
examples on other cars:
![gradcam_cam (3)](https://user-images.githubusercontent.com/55695300/147957241-1da45ccf-30b3-4fb0-b6a0-ade459232e2e.jpg)

even with different aspect ratio:
![Inkedgradcam_cam (4)_LI](https://user-images.githubusercontent.com/55695300/147957917-eaac0a52-8c0d-4012-8e17-cb1338361011.JPG)


same activation map, but all of them are classified correctly

**Comment by jacobgil** [comment]
This is weird. The code looks ok.
I'm thinking how could we debug this.

Just want to make sure that there is no "silly" mistake since the CAMs are the same -
The coding is reading the next CAM every time, right? It's not always taking the first image?

**Comment by jacobgil** [comment]
Is there a way you could share the weights of the model, and the script you're running?
So I can help debug it?

**Comment by davcaste** [comment]
I'm not sure how i can upload the .pt here because it seems not to find it in my folder.

Anyway the problem seems to be in my trained model because by making the only change of commenting the line in which i load the weights that i computing by training the model, the cam that is produced from the last layer of the feature extractor of resnet34 pretrained on imagenet has actually sense.

So, since I'm looking at the last layer of the feature extractor and I retrained the network from scratch, it seems that the feature extractor gives always the same output (maybe I have few data for training all the parameters) for each image while all the work is done by the classification head. 

I will try to retrain just the classification head on top of the feature extractor trained on imagenet and also I will try to produce the activation map of the classification head (if this is possible).

Do you think it can be the right way to go?

**Comment by jacobgil** [comment]
Hi,

The activation maps are only for 2D activations, so it won't work for the classification head (only for the features right before it after the convolutions).

Even if the custom network learned to do the work mostly in the fully connected layers, it's still strange that the feature extractor always gives the same output. The fully connected layers work with the output of the feature extractor, so some information is supposed to be there.
What you suggested would be worth trying to see if the result is different.

If you can upload the .pt file (and an example script to run it) to google drive, drop-box etc.
(I will be able to take a look around Friday).

**Comment by davcaste** [comment]
Hi, 
I found out that the error was in the normalization since I was using mean = [0.485, 0.456, 406] in my trained model instead of [0.485, 0.456, 0.406], once i changed that the visualization works good.
Thank you anyway for your help and sorry for the time you lost for that.
